### ISSUE TITLE
To think about some cool badge?

### ISSUE THREAD

**Issue by charlike** [singapore/renovate#534]
Hey! :hand: 

I'm thinking to drop most of my badges and stick to as minimal as possible, but want to show that my deps are always up to date, because Semantic-Release (probably) and Renovate and good its specific config that i'm thinking now (low noises, big automation, plus pinning versions).

So i think this would be enough
[![Renovate - Always up to date][renovateapp-svg]][renovateapp-link]

[renovateapp-link]: https://renovateapp.com
[renovateapp-svg]: https://img.shields.io/badge/always-up_to_date-brightgreen.svg

```
[![Renovate - Always up to date][renovateapp-svg]][renovateapp-link]

[renovateapp-link]: https://renovateapp.com
[renovateapp-svg]: https://img.shields.io/badge/always-up_to_date-brightgreen.svg
```

**Comment by rarkins** [comment]
My thoughts:
- Installing Renovate doesn't necessarily mean you're up-to-date :) Even with automerge, some tests might fail..  How about ![monitored](https://img.shields.io/badge/dependencies-renovated-brightgreen.svg) or similar?
- Any value in having a dynamic badge that indicates if a repo has any outstanding updates? The downside is that most repos will have *some* dependencies out of date at some time, and then it might be discouraging to people if they always have a yellow or orange colour.
- Maybe a tricolor badge that shows if a repo is up-to-date for major/minor/patch?
- Or are we duplicating https://david-dm.org/ ?

**Comment by rarkins** [comment]
If we were doing statuses:
- "up to date" and bright green if no updates at all outstanding
- "patched" and green if all minor/patch updates applied but any major ones outstanding
- "out of date" and yellow if minor out of date
- "out of date" and red if patch out of date

But perhaps also with some "delay" incorporated? e.g. it's a bit annoying if your repository suddenly turns red the minute a new patch is out and you will have it patched within an hour or few hours or whatever.

**Comment by ikatyang** [comment]
How about indicating if Renovate is enabled for the repo?

[![Renovate](https://img.shields.io/badge/Renovate-enabled-brightgreen.svg)](https://renovateapp.com/)
[![Renovate](https://img.shields.io/badge/Renovate-not_found-lightgray.svg)](https://renovateapp.com/)

**Comment by rarkins** [comment]
Yeah I'm thinking of how to do it in a forwards-compatible way so that it can be enhanced in future without changes to everyone's readme. Maybe I set up and dummy AWS API gateway endpoint that simply answers with "renovate enabled" and then later improve it to:

- show "disabled" if the repo is not known!
- show advanced details about a repo's status (out of date, etc)

Also need to think how to do it for private repos without it being a security concern for those (e.g. Needs a secret strong in addition to repo name)

**Comment by tunnckoCore** [comment]
> How about monitored or similar?

Thought exactly for that, but is too long for my taste (David's "up-to-date" is long too) :D As about the other things, I believe there is a room for such service like David DM - your ideas sounds good.

> answers with "renovate enabled" and then later improve it to:

Probably yea. It's a good start.

**Comment by hutson** [comment]
> Installing Renovate doesn't necessarily mean you're up-to-date

I agree with that statement. While seeing a `renovate` badge will tell me a little bit about the efforts a package manager has put into the sustainability of their project, it doesn't tell me whether their package is, at that moment, up-to-date with their dependencies. They could quite easily leave `renovate` installed, but neglect to actually accept any `renovate`-generated pull requests.

> Or are we duplicating https://david-dm.org/ ?

As a major consumer of npm packages, I actively look at `david-dm` badges on projects, and will take into consideration the state of those badges when deciding whether to use a package.

So maybe both badges are, in their own way, telling me part of the story about the suitability of a package for consumption. One is indicating whether the package is actually up-to-date, while the other tells me how much the maintainer cares about their package, and if the package is currently out-of-date, that perhaps it will be auto-updated in the near future.

**Comment by rarkins** [comment]
Hi all, thanks for the feedback. Based on your inputs, my current line of thinking is:

- David DM badge is useful/widespread and no need to duplicate it for now
- Whether a repository is up to date and whether it has renovate installed are separate matters :)
- For now, a simple voluntary and static "renovate enabled" badge should be fine
- Later, I will consider adding a dynamic badge for the service itself, which can indicate:
   * enabled or disabled
   * Number of PRs outstanding

So I think that the combination of "are dependencies up to date" courtesy of David, plus "number of PRs outstanding" could be good and complimentary.

So for example if someone is on Webpack 2.x and has decided not to upgrade to Webpack 3.x (i.e. closed the Renovate PR), then it might show as slightly out of date on David, but would show as green/0 PRs outstanding on renovate.

**Comment by daKmoR** [comment]
As seen here I was looking at david dm badge - and that really works well for "normal" git repositories. E.g. if you have one package.json file in the root of your git.

However my use case is a monorepo and I really like renovate because it picks up all the nested package.json and handles them.

So on how it behaves and looks, I would say david dm is really nice... so having the same feature but aggregated for ALL package.json files would be awesome :)

My guess would be that something like this would be out of scope for david dm*, but renovate should already have all the needed data right?

\* I asked let's see what they say.

**Comment by rarkins** [comment]
Thanks for the update. It's useful to know if there's a feature gap in David DM now with monorepos.

For me the challenge is (1) how to "fairly" summarise update statuses into some single score, and (2) how to fit it into a single badge. e.g. a patch update open for 2 months is a lot different to a major upgrade open for the same. 

Although if we do it with an API, I suppose that it's possible to support a few different approaches and let people choose which they use.

**Comment by nottrobin** [comment]
I'd also appreciate it if Renovate included a badge that could summarise the status of Python dependencies too. Or one that shows the status of all dependencies.

**Comment by tunnckoCore** [comment]
Yea, one more reason why Renovate should have a badge, and that isn't a duplication of DavidDM. Now Renovate supports more than NPM/nodejs so a badge showing a status would be great. Renovate will become DavidDM for all the languages that it supports :D

**Comment by rarkins** [comment]
OK, I plan to add this soon. To begin with, the badge will just say onboarding/enabled/disabled, but later on (without change of badge URL) will be updated to be more dynamic and perhaps even customisable, e.g. to display number of dependencies out of date, number of PRs, a score, etc.

**Comment by tunnckoCore** [comment]
@rarkins yup, sounds good. 

You can try to use [npm.im/badgen](https://www.npmjs.com/package/badgen) library (not the badgen service). :)

**Comment by rarkins** [comment]
Please help me through the logic here. I want it to be able to work both for self-hosted CLI users as well as the app, even though the app one is intended to be much more sophisticated.

So:

New config option `badgeUrl` which defaults to `null`. A PR proposing the badge will be created only if it's a non-null value, i.e. by default there is no Renovate badge in the CLI tool. CLI users could simply set it to a static SVG URL that says "renovate enabled".

If `badgeUrl` is non-null and the repo contains a readme.md then a PR will be raised only if the repository is already onboarded. i.e. for new repos, the PR will come after onboarding is completed but for existing repos they will all get a PR at the same time once this feature is turned on in the app.

To ignore/decline the badge suggestion:
- Ideally, you just close the PR and Renovate can find it again. Possibly though if you end up with 1000+ closed Renovate PRs then the badge comes back again because GitHub doesn't tell us about the closed PR anymore
- You can fully decline it by setting `"badgeUrl": null` in renovate config, including preset. This will stop it ever being created too, if present

The app will set the image to a dynamic renovatebot.com URL. Initially this will say "enabled", "disabled", "onboarding" or "unknown". The image will link to the dashboard for the repo.

In future, how to make it customisable? In the best tradition of Renovate, I think it should be via config, e.g. a new option `badgeParameters` or something like that. Renovate CLI would ignore them, but Renovate the App would read these during its normal run and write them to its backend DB and then use that to determine what to display whenever the badge is requested. e.g. it could be "simple" (enabled/disabled/etc), it could instead say "7 updates" (meaning 7 dependencies), or it could say "2 PRs" if there is grouping. I guess it's OK to defer this decision.

**Comment by nottrobin** [comment]
Does the badge need to be something managed by Renovate through config?

I understand that filing PRs is what Renovate does, but I'd personally feel more comfortable if it only did that specifically for its core MO of upgrading dependencies, and not for all sorts of other things like adding badges. Otherwise I'd start to feel like I didn't know what Renovate might do on my repo next.

Like, what happens if Renovate files the PR to add the badge, it's merged, and then you set `badgeUrl: null` again - does it now file a PR to remove it again from the README? What if it's been modified? I feel it would make Renovate users feel more empowered if they added that to the README.md themselves.

All I was expecting was that somewhere in documentation it would tell me how to add a renovate badge to my README.md myself, with a snippet similar to:

``` markdown
![renovate dependency status](https://renovatebot.com/badges/dependency-status.svg?repo=canonical-webteam/docs.snapcraft.io&branch=master)
```

(Any additional options could just be supported as further query parameters on the URL. Or through different endpoints.)

In any event, it seems that simply providing a badge image endpoint would be the first step, and then you could explore automation options for managing it through config as an enhancement.

**Comment by nottrobin** [comment]
Those are all very good points. It is especially true that people are much more likely to add it under your model. I agree that a separate PR is better.

I think, though, for that purpose, you might want to simplify the config option to just `badge: true|false`? Presumably Renovate is the one that knows the URL, and probably wants the ability to change it. In future, you could extend it to `badge: { ... options ... }`, when there are further options.

**Comment by tunnckoCore** [comment]
Absolutely agree with @nottrobin. It sounds more complicated than it should be. Just expose an API endpoint that gives you status metadata as json, so we can use it. And another endpoint for the badge.

> I think, though, for that purpose, you might want to simplify the config option to just 

Yea.. I don't get what `badgeUrl` could be. `true/false` sounds me better.

> @rarkins: What I can maybe do though is make a random selection of users accounts initially

Yea, definitely better :D

And yes, for sure, it's more likely most will accept those PRs than consider adding a badge themselves manually. But I'm against such PRs because have strict preferences about my readmes and I even don't modify them manually, I generate them. Even if didn't do that, still don't like the idea because no one can know where to put that badge exactly. Yea, there is some convention, but still.

**Comment by rarkins** [comment]
@tunnckoCore I think you're potentially the only user who autogenerates their readme's, so I don't want to optimise for that *particular* case. i.e. so long as you can disable automatic PR insertion, that's all you need.

`badgeUrl` would only be the concern of bot admins, not of bot users.

For the app's URL, I would prefer to keep its URL simple and control its contents via config. That way - regardless of whether the badge is added with automated PRs or manually - it means that users can centrally control (via presets) how their badges look, which will be particularly important assuming the badge's richness and configurability will be improved over time (e.g. including changing colour if there are vulnerabilities among the updates waiting).

Right now the URL in planning would be like https://badges.renovateapi.com/github/renovateapp/renovate

**Comment by cdaringe** [comment]
[![Renovate](https://img.shields.io/badge/renovate-enabled-brightgreen.svg)](https://renovatebot.com)

`[![Renovate](https://img.shields.io/badge/renovate-enabled-brightgreen.svg)](https://renovatebot.com)`

for now, i'm using shields.io ^

**Comment by JamieMagee** [comment]
@cdaringe Nice! I took your idea a little further, and added the renovate logo. I put the code [in a gist](https://gist.github.com/JamieMagee/63feb6f16e0acb8f4624a0310a413661)

[![Renovate](https://img.shields.io/badge/renovate-enabled-green?style=for-the-badge&logo=data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCAzNjkgMzY5Ij48Y2lyY2xlIGN4PSIxODkuOSIgY3k9IjE5MC4yIiByPSIxODQuNSIgZmlsbD0iI2ZmZTQyZSIgdHJhbnNmb3JtPSJ0cmFuc2xhdGUoLTUgLTYpIi8+PHBhdGggZmlsbD0iIzhiYjViNSIgZD0iTTI1MSAyNTZsLTM4LTM4YTE3IDE3IDAgMDEwLTI0bDU2LTU2YzItMiAyLTYgMC03bC0yMC0yMWE1IDUgMCAwMC03IDBsLTEzIDEyLTktOCAxMy0xM2ExNyAxNyAwIDAxMjQgMGwyMSAyMWM3IDcgNyAxNyAwIDI0bC01NiA1N2E1IDUgMCAwMDAgN2wzOCAzOHoiLz48cGF0aCBmaWxsPSIjZDk1NjEyIiBkPSJNMzAwIDI4OGwtOCA4Yy00IDQtMTEgNC0xNiAwbC00Ni00NmMtNS01LTUtMTIgMC0xNmw4LThjNC00IDExLTQgMTUgMGw0NyA0N2M0IDQgNCAxMSAwIDE1eiIvPjxwYXRoIGZpbGw9IiMyNGJmYmUiIGQ9Ik04MSAxODVsMTgtMTggMTggMTgtMTggMTh6Ii8+PHBhdGggZmlsbD0iIzI1YzRjMyIgZD0iTTIyMCAxMDBsMjMgMjNjNCA0IDQgMTEgMCAxNkwxNDIgMjQwYy00IDQtMTEgNC0xNSAwbC0yNC0yNGMtNC00LTQtMTEgMC0xNWwxMDEtMTAxYzUtNSAxMi01IDE2IDB6Ii8+PHBhdGggZmlsbD0iIzFkZGVkZCIgZD0iTTk5IDE2N2wxOC0xOCAxOCAxOC0xOCAxOHoiLz48cGF0aCBmaWxsPSIjMDBhZmIzIiBkPSJNMjMwIDExMGwxMyAxM2M0IDQgNCAxMSAwIDE2TDE0MiAyNDBjLTQgNC0xMSA0LTE1IDBsLTEzLTEzYzQgNCAxMSA0IDE1IDBsMTAxLTEwMWM1LTUgNS0xMSAwLTE2eiIvPjxwYXRoIGZpbGw9IiMyNGJmYmUiIGQ9Ik0xMTYgMTQ5bDE4LTE4IDE4IDE4LTE4IDE4eiIvPjxwYXRoIGZpbGw9IiMxZGRlZGQiIGQ9Ik0xMzQgMTMxbDE4LTE4IDE4IDE4LTE4IDE4eiIvPjxwYXRoIGZpbGw9IiMxYmNmY2UiIGQ9Ik0xNTIgMTEzbDE4LTE4IDE4IDE4LTE4IDE4eiIvPjxwYXRoIGZpbGw9IiMyNGJmYmUiIGQ9Ik0xNzAgOTVsMTgtMTggMTggMTgtMTggMTh6Ii8+PHBhdGggZmlsbD0iIzFiY2ZjZSIgZD0iTTYzIDE2N2wxOC0xOCAxOCAxOC0xOCAxOHpNOTggMTMxbDE4LTE4IDE4IDE4LTE4IDE4eiIvPjxwYXRoIGZpbGw9IiMzNGVkZWIiIGQ9Ik0xMzQgOTVsMTgtMTggMTggMTgtMTggMTh6Ii8+PHBhdGggZmlsbD0iIzFiY2ZjZSIgZD0iTTE1MyA3OGwxOC0xOCAxOCAxOC0xOCAxOHoiLz48cGF0aCBmaWxsPSIjMzRlZGViIiBkPSJNODAgMTEzbDE4LTE3IDE4IDE3LTE4IDE4ek0xMzUgNjBsMTgtMTggMTggMTgtMTggMTh6Ii8+PHBhdGggZmlsbD0iIzk4ZWRlYiIgZD0iTTI3IDEzMWwxOC0xOCAxOCAxOC0xOCAxOHoiLz48cGF0aCBmaWxsPSIjYjUzZTAyIiBkPSJNMjg1IDI1OGw3IDdjNCA0IDQgMTEgMCAxNWwtOCA4Yy00IDQtMTEgNC0xNiAwbC02LTdjNCA1IDExIDUgMTUgMGw4LTdjNC01IDQtMTIgMC0xNnoiLz48cGF0aCBmaWxsPSIjOThlZGViIiBkPSJNODEgNzhsMTgtMTggMTggMTgtMTggMTh6Ii8+PHBhdGggZmlsbD0iIzAwYTNhMiIgZD0iTTIzNSAxMTVsOCA4YzQgNCA0IDExIDAgMTZMMTQyIDI0MGMtNCA0LTExIDQtMTUgMGwtOS05YzUgNSAxMiA1IDE2IDBsMTAxLTEwMWM0LTQgNC0xMSAwLTE1eiIvPjxwYXRoIGZpbGw9IiMzOWQ5ZDgiIGQ9Ik0yMjggMTA4bC04LThjLTQtNS0xMS01LTE2IDBMMTAzIDIwMWMtNCA0LTQgMTEgMCAxNWw4IDhjLTQtNC00LTExIDAtMTVsMTAxLTEwMWM1LTQgMTItNCAxNiAweiIvPjxwYXRoIGZpbGw9IiNhMzM5MDQiIGQ9Ik0yOTEgMjY0bDggOGM0IDQgNCAxMSAwIDE2bC04IDdjLTQgNS0xMSA1LTE1IDBsLTktOGM1IDUgMTIgNSAxNiAwbDgtOGM0LTQgNC0xMSAwLTE1eiIvPjxwYXRoIGZpbGw9IiNlYjZlMmQiIGQ9Ik0yNjAgMjMzbC00LTRjLTYtNi0xNy02LTIzIDAtNyA3LTcgMTcgMCAyNGw0IDRjLTQtNS00LTExIDAtMTZsOC04YzQtNCAxMS00IDE1IDB6Ii8+PHBhdGggZmlsbD0iIzEzYWNiZCIgZD0iTTEzNCAyNDhjLTQgMC04LTItMTEtNWwtMjMtMjNhMTYgMTYgMCAwMTAtMjNMMjAxIDk2YTE2IDE2IDAgMDEyMiAwbDI0IDI0YzYgNiA2IDE2IDAgMjJMMTQ2IDI0M2MtMyAzLTcgNS0xMiA1em03OC0xNDdsLTQgMi0xMDEgMTAxYTYgNiAwIDAwMCA5bDIzIDIzYTYgNiAwIDAwOSAwbDEwMS0xMDFhNiA2IDAgMDAwLTlsLTI0LTIzLTQtMnoiLz48cGF0aCBmaWxsPSIjYmY0NDA0IiBkPSJNMjg0IDMwNGMtNCAwLTgtMS0xMS00bC00Ny00N2MtNi02LTYtMTYgMC0yMmw4LThjNi02IDE2LTYgMjIgMGw0NyA0NmM2IDcgNiAxNyAwIDIzbC04IDhjLTMgMy03IDQtMTEgNHptLTM5LTc2Yy0xIDAtMyAwLTQgMmwtOCA3Yy0yIDMtMiA3IDAgOWw0NyA0N2E2IDYgMCAwMDkgMGw3LThjMy0yIDMtNiAwLTlsLTQ2LTQ2Yy0yLTItMy0yLTUtMnoiLz48L3N2Zz4=)](https://renovatebot.com)

**Comment by vidavidorra** [comment]
Would love to see this feature~ @rarkins is there any update/timeline on it?
